Include/pymacro.h

```diff
@@ -3,20 +3,20 @@
 
 // gh-91782: On FreeBSD 12, if the _POSIX_C_SOURCE and _XOPEN_SOURCE macros are
 // defined, <sys/cdefs.h> disables C11 support and <assert.h> does not define
-// the static_assert() macro. Define the static_assert() macro in Python until
-// <sys/cdefs.h> suports C11:
+// the static_assert() macro.
 // https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=255290
-#if defined(__FreeBSD__) && !defined(static_assert)
-#  define static_assert _Static_assert
-#endif
-
-// static_assert is defined in glibc from version 2.16. Before it requires
-// compiler support (gcc >= 4.6) and is called _Static_assert.
+//
+// macOS <= 10.10 doesn't define static_assert in assert.h at all despite
+// having C11 compiler support.
+//
+// static_assert is defined in glibc from version 2.16. Compiler support for
+// the C11 _Static_assert keyword is in gcc >= 4.6.
+//
+// MSVC makes static_assert a keyword, contrary to the C standard.
+//
 // In C++ 11 static_assert is a keyword, redefining is undefined behaviour.
-#if (defined(__GLIBC__) \
-     && (__GLIBC__ < 2 || (__GLIBC__ == 2 && __GLIBC_MINOR__ <= 16)) \
-     && !(defined(__cplusplus) && __cplusplus >= 201103L) \
-     && !defined(static_assert))
+#if !defined(static_assert) && !defined(_MSC_VER) \
+     && !(defined(__cplusplus) && __cplusplus >= 201103L)
 #  define static_assert _Static_assert
```

Comment:
Clang will need also a check for `__has_feature(c_static_assert)` here because it may be that `_Static_assert` is not defined

Reply:
Is that really necessary given that C11 is required and  `_Static_assert` is part of that standard? Or is the concern that C99 projects might try to use `static_assert` if it's defined?

Reply:
> In practice, which platform/libc/compiler is impacted by this change?

I only know of the ones mentioned in the comment, i.e. FreeBSD <= 12, macOS <= 10.10, and glibc < 2.16.

Reply:
Some compiler versions don't implement the full standard (Microsoft compiler is the obvious one but there are more) and some old versions of clang / libc may not implement it either. I also don't know what clang does on AIX for instance, where half of these things are just missing

Reply:
AIX is not longer offiically supported by Python: https://peps.python.org/pep-0011/ If someone gets an issue on AIX and proposes a PR, I'm open to review it. But I prefer to not make assumptions on how AIX should or should not behave.

Reply:
Does it work to express the test as ``!(defined(__has_feature) && __has_feature(c_static_assert))``? If no, maybe we should consider adding _Py__has_feature(), based on this existing macro:

```c
// Preprocessor check for a builtin preprocessor function. Always return 0
// if __has_builtin() macro is not defined.
//
// __has_builtin() is available on clang and GCC 10.
#ifdef __has_builtin
#  define _Py__has_builtin(x) __has_builtin(x)
#else
#  define _Py__has_builtin(x) 0
#endif
```

If I recall correctly, it doesn't work to write ``#if defined(__has_builtin) && __has_builtin(xxx)``.

Reply:
`_Static_assert` is a language keyword, so libc shouldn't be a factor, just the compiler. Would replacing the `__cplusplus` checks with `defined(__STDC_VERSION__) && __STDC_VERSION__ >= 201112L` be sufficient? Seems like the worst that could happen in that case is you go from an error saying "static_assert undefined" to one saying "_Static_assert undefined".

Reply:
> Does it work to express the test as `!(defined(__has_feature) && __has_feature(c_static_assert))`?

It wouldn't, because the preprocessor evaluates all macros on a line before doing anything else, so short-circuit evaluation doesn't prevent it from trying to evaluate the possibly-undefined thing on the right.

Reply:
> AIX is not longer offiically supported by Python: https://peps.python.org/pep-0011/ If someone gets an issue on AIX and proposes a PR, I'm open to review it. But I prefer to not make assumptions on how AIX should or should not behave.

I know, that was just an example